practical/ProcessMining/group2/conformance_checking/src/generate_footprint.py

```diff
@@ -0,0 +1,177 @@
+from itertools import chain
+import numpy as np
+from enum import Enum
+from natsort import natsorted
+
+
+class Relations(Enum):
+    BEFORE = '<-'
+    SEQUENCE = '->'
+    NOT_FOLLOWED = '#'
+    PARALLEL = '||'
+
+
+class FootPrintMatrix:
+
+    def __init__(self, log=None, relations=None):
+        """
+        Initialize a FootPrintMatrix instance.
+
+        Parameters:
+        ----------
+        log : dict, optional
+            The input log to be converted into a list of traces for the footprint matrix.
+        relations : dict, optional
+            A dictionary representing the relations in the footprint matrix.
+        """
+        if log is not None:
+            self.traces = self.convert_log_for_footprintmatrix(log)
+        else:
+            self.traces = []
+        self.transitions = set()
+
+        if relations is None:
+            self.relations = {}
+        else:
+            self.relations = relations
+
+        self.places = []
+
+    @classmethod
+    def from_relations(cls, relations):
+        """
+        Create a FootPrintMatrix instance from a given relations dictionary.
+
+        This class method initializes a FootPrintMatrix object using the provided
+        relations dictionary.
+
+        Parameters:
+        ----------
+        relations : dict
+            The relations dictionary to initialize the FootPrintMatrix.
+
+        Returns:
+        -------
+        FootPrintMatrix
+            A new FootPrintMatrix instance with the specified relations.
+        """
+        return cls(relations=relations)
+
+    def sort_fpm_rec(self, relations):
```

Comment:
contemplate renaming to sort_footprint_matrix_recursively, criptic method title's never solved anything ^^